README.md
Path MTU Discovery Daemon
===============

## Overview

Devices behind VPN or other weird network links are unable to use [Path MTU Discovery](https://tools.ietf.org/html/rfc1191). This causes undesired fragmentation on the physical link between the device and the VPN endpoint.
If we know (i.e. calculate) the overhead added by the encapsulation of the prococol we use, we can directly listen to IPv4 packets with the DF (Don't Forward) bit set and send back ICMP type 3 code 4 messages with the correct MTU.
That is exaclty what PMTUDD (Path MTU Discovery Daemon) does.

Note: The source code for this project is derived from CloudFlare's [pmtud](https://github.com/cloudflare/pmtud), according to the project's license. I am incredibly thankful to have found this codebase from which I could build upon.

## Usage

```
$ ./pmtudd --help

Usage:

    pmtudd [options]

Path MTU Discovery Daemon listens for inbound IPv4 packets with the DF (Don't Fragment)
bit set and sends back ICMP code 3 messages related to MTU detection

Options:

  --iface              Network interface to listen on
  --src-rate           Pps limit from single source (default=1.1 pss)
  --iface-rate         Pps limit to send on a single interface (default=10.0 pps)
  --verbose            Print forwarded packets on screen
  --dry-run            Don't inject packets, just dry run
  --desired-mtu        The MTU to send back link
  --cpu                Pin to particular cpu
  --help               Print this message

Example:

    pmtudd --iface=eth2 --src-rate=1.1 --iface-rate=10.0 --desired-mtu=1420

```

## Building

    git submodule update --init --recursive
    make

## Additional information

pmtudd uses the following simple filter to match the approriate IPv4 packets:

    ip and ip[6] == 64 and greater {{MTU}}

It then filters out possibly invalid/bogus packets and ignores packets which have the same src and dst MAC addresses (this prevents an attacker from performing DoS attacks).
If the packet is valid, pmtudd crafts a new ICMP type 3 code 4 packet and sends it back to the address it received the original IPv4 packet from.

## Use case

This was developped as a reponse to the issue of running hosts behind a VXLAN over Wireguard tunnel. The hosts used a default MTU of 1500 and were unable to detect the appropriate MTU to use.
Using iptable rules on the wireless router to which they were connected in order to set the TCP MSS value was obviously not a good enough option and leaving all packets getting fragmented once sent on the physical link had a huge performance impact.

This could be useful in all situations where devices connect to a network which forwards all packets through a VPN tunnel (be it IPSec, Wireguard, OpenVPN or any other) and where the physical MTU of the underlying network cannot be modified.

RELEASE_NOTES
v0.7  (20191104-0225)
- 2016-01-08 Fix #1 - typo: called nflog_unbind_pf instead of nflog_bind_pf
- 2016-01-20 Only capture incoming packets
- 2018-09-01 fix --strict option crash
- 2018-09-10 Merge pull request #7 from kerolasa/crash-fix
- 2018-09-11 avoid buffer overflow
- 2018-09-12 Merge pull request #9 from kerolasa/fix-possible-overflow
- 2018-10-26 Add option to use src MAC of orig pkt as a src of fwd pkt
- 2018-10-26 Fix help text formatting
- 2019-05-03 Merge pull request #10 from niku64/ft-use-src-mac
- 2019-11-02 Repurpose completely the project to fit our needs
- 2019-11-02 Rename to pmtudd and update README
- 2019-11-04 Bring bag RELEASE_NOTES
- 2019-11-04 Remove unneeded dependencies and cleanup Makefile

v0.6  (20150219-1309)
- 2015-02-18 Tweak csiphash to eliminate misaligned reads
- 2015-02-18 Split the hashlimit into two parts: update and subtraction
- 2015-02-18 Update again SRC_RATE_PPS, this time reduce to 1.1
- 2015-02-19 Don't forward messages with MTU of next hop == 0
- 2015-02-19 nflog: dependencies
- 2015-02-19 nflog: support --nflog option, as alternative to pcap for receiving
- 2015-02-19 cosmetic: make format
- 2015-02-19 Makefile: fix missing build dependencies
- 2015-02-19 nflog: make sure the messages aren't buffered for default 1 second
- 2015-02-19 Cosmetic: ignore debs during make

v0.5  (20150218-1754)
- 2015-02-18 Fix bug with wrong offsets in parsing ICMP payloads
- 2015-02-18 Bump --src-pps default to 1.9 pps

v0.4  (20150218-1504)
- 2015-02-18 Fix crash due to inappropriate memory allocation in bitmap

v0.3  (20150218-1259)
- 2015-02-16 Forward packets coming out only from supplied outgoing ports

v0.2  (20150212-1720)
- 2015-02-09 Ignore ENOBUFS on send()
- 2015-02-12 Add --cpu option to allow pmtud to be pinned to a cpu
- 2015-02-12 Clearer readme and --help
- 2015-02-12 Update build process

v0.1  (20150131-0037)


src/net.c
// PMTUD
//
// Copyright (c) 2015 CloudFlare, Inc.

#include <getopt.h>
#include <pcap.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <sys/socket.h>
#include <arpa/inet.h>
#include <linux/if_packet.h>
#include <net/ethernet.h>
#include <net/if.h>
#include <string.h>
#include <sys/ioctl.h>

#include "pmtudd.h"

pcap_t *setup_pcap(const char *iface, const char *bpf_filter, int snap_len,
		   struct pcap_stat *stats)
{

	char errbuf[PCAP_ERRBUF_SIZE] = {0};
	pcap_t *pcap = pcap_create(iface, errbuf);
	if (pcap == NULL) {
		FATAL("pcap_create(%s): %s", str_quote(iface), errbuf);
	}

	int r;

	r = pcap_set_tstamp_type(pcap, PCAP_TSTAMP_ADAPTER_UNSYNCED);
	if (r != 0) {
		pcap_set_tstamp_type(pcap, PCAP_TSTAMP_ADAPTER);
	}

	r = pcap_set_promisc(pcap, 0);
	if (r != 0) {
		FATAL("pcap_set_promisc: %s", pcap_geterr(pcap));
	}

	r = pcap_set_snaplen(pcap, snap_len);
	if (r != 0) {
		FATAL("pcap_set_snaplen: %s", pcap_geterr(pcap));
	}

	r = pcap_set_timeout(pcap, 100 /* ms */);
	if (r != 0) {
		FATAL("pcap_set_timeout: %s", pcap_geterr(pcap));
	}

	r = pcap_activate(pcap);
	if (r != 0) {
		FATAL("pcap_activate(%s): %s", str_quote(iface),
		      pcap_geterr(pcap));
	}

	r = pcap_setdirection(pcap, PCAP_D_IN);
	if (r != 0) {
		FATAL("pcap_setdirection: %s %d", pcap_geterr(pcap), r);
	}

	r = pcap_setnonblock(pcap, 1, errbuf);
	if (r != 0) {
		FATAL("pcap_setnonblock(%s, 1): %s", str_quote(iface), errbuf);
	}

	r = pcap_getnonblock(pcap, errbuf);
	if (r != 1) {
		FATAL("pcap_getnonoblock(): can't set nonblocking!");
	}

	struct bpf_program bpf;

	r = pcap_compile(pcap, &bpf, bpf_filter, 1, PCAP_NETMASK_UNKNOWN);
	if (r != 0) {
		FATAL("pcap_compile(%s): %s", str_quote(bpf_filter),
		      pcap_geterr(pcap));
	}

	r = pcap_setfilter(pcap, &bpf);
	if (r != 0) {
		FATAL("pcap_setfilter(%s): %s", str_quote(bpf_filter),
		      pcap_geterr(pcap));
	}

	pcap_freecode(&bpf);

	r = pcap_datalink(pcap);
	if (r != DLT_EN10MB) {
		FATAL("Only Ethernet devices supported %s is %s",
		      str_quote(iface), pcap_datalink_val_to_name(r));
	}

	r = pcap_stats(pcap, stats);
	if (r != 0) {
		FATAL("pcap_stats(%s): %s", str_quote(iface),
		      pcap_geterr(pcap));
	}

	return pcap;
}

void unsetup_pcap(pcap_t *pcap, const char *iface, struct pcap_stat *stats)
{
	struct pcap_stat stats2 = {0, 0, 0};
	int r = pcap_stats(pcap, &stats2);
	if (r != 0) {
		FATAL("pcap_stats(%s): %s", str_quote(iface),
		      pcap_geterr(pcap));
	}

	stats->ps_recv = stats2.ps_recv - stats->ps_recv;
	stats->ps_drop = stats2.ps_drop - stats->ps_drop;
	stats->ps_ifdrop = stats2.ps_ifdrop - stats->ps_ifdrop;

	pcap_close(pcap);
}

int setup_raw(const char *iface)
{
	int r;
	int s = socket(PF_PACKET, SOCK_RAW, htons(ETH_P_ALL));
	if (s < 0) {
		PFATAL("socket(PF_PACKET, SOCK_RAW, htons(ETH_P_ALL))");
	}

	struct ifreq s_ifr;
	memset(&s_ifr, 0, sizeof(s_ifr));
	strncpy(s_ifr.ifr_name, iface, sizeof(s_ifr.ifr_name) - 1);
	r = ioctl(s, SIOCGIFINDEX, &s_ifr);
	if (r != 0) {
		PFATAL("ioctl(SIOCGIFINDEX, %s)", str_quote(iface));
	}

	struct sockaddr_ll my_addr;
	memset(&my_addr, 0, sizeof(my_addr));
	my_addr.sll_family = AF_PACKET;
	my_addr.sll_protocol = htons(ETH_P_ALL);
	my_addr.sll_ifindex = s_ifr.ifr_ifindex;
	r = bind(s, (struct sockaddr *)&my_addr, sizeof(my_addr));
	if (r != 0) {
		PFATAL("bind(%s, AF_PACKET)", str_quote(iface));
	}

	int tmp = 1024 * 1024;
	r = setsockopt(s, SOL_SOCKET, SO_SNDBUF, &tmp, sizeof(tmp));
	if (r != 0) {
		PFATAL("setsockopt(SO_SNDBUF)");
	}

	/* r = fcntl(s, F_SETFL, */
	/* 	  O_NONBLOCK | fcntl(s, F_GETFL, 0)); */
	/* if (r != 0) { */
	/* 	PFATAL("fcntl(O_NONBLOCK)"); */
	/* } */
	return s;
}

const char *ip_to_string(const uint8_t *p, int p_len)
{
	static char dst[INET6_ADDRSTRLEN + 1];
	const char *r = NULL;

	if (p_len == 4) {
		struct in_addr addr;
		memcpy(&addr, p, 4);
		r = inet_ntop(AF_INET, &addr, dst, INET6_ADDRSTRLEN);
	}
	if (p_len == 16) {
		struct in6_addr addr;
		memcpy(&addr, p, 16);
		r = inet_ntop(AF_INET6, &addr, dst, INET6_ADDRSTRLEN);
	}

	if (r == NULL) {
		dst[0] = '?';
		dst[1] = 0x00;
	}
	return dst;
}


src/pmtud.h
// PMTUD
//
// Copyright (c) 2015 CloudFlare, Inc.

#define ERRORF(x...) fprintf(stderr, x)

#define FATAL(x...)                                                            \
	do {                                                                   \
		ERRORF("[-] PROGRAM ABORT : " x);                              \
		ERRORF("\n\tLocation : %s(), %s:%u\n\n", __FUNCTION__,         \
		       __FILE__, __LINE__);                                    \
		exit(EXIT_FAILURE);                                            \
	} while (0)

#define PFATAL(x...)                                                           \
	do {                                                                   \
		ERRORF("[-] SYSTEM ERROR : " x);                               \
		ERRORF("\n\tLocation : %s(), %s:%u\n", __FUNCTION__, __FILE__, \
		       __LINE__);                                              \
		perror("      OS message ");                                   \
		ERRORF("\n");                                                  \
		exit(EXIT_FAILURE);                                            \
	} while (0)

#define TIMESPEC_NSEC(ts) ((ts)->tv_sec * 1000000000ULL + (ts)->tv_nsec)

#define NSEC_TIMEVAL(ns)                                                       \
	(struct timeval)                                                       \
	{                                                                      \
		(ns) / 1000000000ULL, ((ns) % 1000000000ULL) / 1000ULL         \
	}
#define MSEC_NSEC(ms) ((ms)*1000000ULL)

struct icmp_packet
{
    uint8_t eth_dst[8];
    uint8_t eth_src[8];
    uint16_t eth_type;
    uint8_t ip_v_ihl,
            ip_tos;
    uint16_t    ip_length;
    uint16_t    ip_id;
    uint16_t   ip_flags_offset;
    uint8_t ip_ttl,
            ip_proto;
    uint16_t ip_chksum;
    uint8_t ip_src_addr[4],
            ip_dst_addr[4];
    uint8_t icmp_type,
            icmp_code;
    uint16_t    icmp_chksum,
            icmp_unused,
            icmp_mtu;
    uint8_t icmp_data[24 + 8]; // Maximum ip header length is 24 bytes
};

/* utils.c */
const char *optstring_from_long_options(const struct option *opt);
uint16_t add_checksum(const u_int8_t * buf, int length);
void build_packet(const struct icmp_packet * packet, u_int8_t **buf, int ip_length, int packet_length);
int set_core_dump(int enable);
const char *str_quote(const char *s);
const char *to_hex(const uint8_t *s, int len);
int signal_desc(int signal);

/* pcap.c */
pcap_t *setup_pcap(const char *iface, const char *bpf_filter, int snap_len,
		   struct pcap_stat *stats);
void unsetup_pcap(pcap_t *pcap, const char *iface, struct pcap_stat *stats);
int setup_raw(const char *iface);
const char *ip_to_string(const uint8_t *p, int p_len);

/* sched.c */
int taskset(int taskset_cpu);

src/utils.c
// PMTUD
//
// Copyright (c) 2015 CloudFlare, Inc.

#include <getopt.h>
#include <pcap.h>
#include <signal.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/resource.h>
#include <sys/signalfd.h>
#include <sys/time.h>

#include "pmtudd.h"

const char *optstring_from_long_options(const struct option *opt)
{
	static char optstring[256] = {0};
	char *osp = optstring;

	for (; opt->name != NULL; opt++) {
		if (opt->flag == 0 && opt->val > 0 && opt->val < 256) {
			*osp++ = opt->val;
			switch (opt->has_arg) {
			case optional_argument:
				*osp++ = ':';
				*osp++ = ':';
				break;
			case required_argument:
				*osp++ = ':';
				break;
			}
		}
	}
	*osp++ = '\0';

	if (osp - optstring >= (int)sizeof(optstring)) {
		abort();
	}

	return optstring;
}

/**
  Returns the sum of the pairs of 2 bytes.
 **/
uint16_t add_checksum(const u_int8_t * buf, int length) {

    int i;
    uint16_t sum = 0, bytes;
    for(i = 1; i< length; i += 2) {
        bytes = (u_int8_t)buf[i] | ((u_int8_t)(buf[i - 1]) << 8);
        sum += bytes;
    }
    return sum;
}

/* @TODO : This is absolutely UGLY */
void build_packet(const struct icmp_packet * packet, u_int8_t **buf, int ip_payload_length, int packet_length) {
    *buf = malloc(sizeof(u_int8_t) * packet_length);
    int cursor = 0;

    memcpy(*buf, packet->eth_dst, 6);
    cursor += 6;

    memcpy(&(*buf)[cursor], packet->eth_src, 6);
    cursor += 6;

    memcpy(&(*buf)[cursor], &packet->eth_type, 2);
    cursor += 2;

    memcpy(&(*buf)[cursor], &packet->ip_v_ihl, 1);
    cursor += 1;

    memcpy(&(*buf)[cursor], &packet->ip_tos, 1);
    cursor += 1;

    memcpy(&(*buf)[cursor], &packet->ip_length, 2);
    cursor += 2;

    memcpy(&(*buf)[cursor], &packet->ip_id, 2);
    cursor += 2;

    memcpy(&(*buf)[cursor], &packet->ip_flags_offset, 2);
    cursor += 2;

    memcpy(&(*buf)[cursor], &packet->ip_ttl, 1);
    cursor += 1;

    memcpy(&(*buf)[cursor], &packet->ip_proto, 1);
    cursor += 1;

    memcpy(&(*buf)[cursor], &packet->ip_chksum, 2);
    cursor += 2;

    memcpy(&(*buf)[cursor], packet->ip_src_addr, 4);
    cursor += 4;

    memcpy(&(*buf)[cursor], packet->ip_dst_addr, 4);
    cursor += 4;

    memcpy(&(*buf)[cursor], &packet->icmp_type, 1);
    cursor += 1;

    memcpy(&(*buf)[cursor], &packet->icmp_code, 1);
    cursor += 1;

    memcpy(&(*buf)[cursor], &packet->icmp_chksum, 2);
    cursor += 2;

    memcpy(&(*buf)[cursor], &packet->icmp_unused, 2);
    cursor += 2;

    memcpy(&(*buf)[cursor], &packet->icmp_mtu, 2);
    cursor += 2;

    memcpy(&(*buf)[cursor], packet->icmp_data, ip_payload_length);
}

int set_core_dump(int enable)
{
	struct rlimit limit;
	limit.rlim_cur = limit.rlim_max = 0;
	if (enable) {
		limit.rlim_cur = limit.rlim_max = RLIM_INFINITY;
	}
	return setrlimit(RLIMIT_CORE, &limit);
}

const char *str_quote(const char *s)
{
	static char buf[1024];
	int r = snprintf(buf, sizeof(buf), "\"%.*s\"", (int)sizeof(buf) - 4, s);
	if (r >= (int)sizeof(buf)) {
		buf[sizeof(buf) - 1] = 0;
	}
	return buf;
}

const char *HEX_CHARS = "0123456789abcdef";

const char *to_hex(const uint8_t *s, int len)
{

	static char buf[1024 + 2];
	if (len > 512) {
		len = 512;
	}

	char *p = buf;
	int i;
	for (i = 0; i < len; i++) {
		p[i * 2] = HEX_CHARS[s[i] >> 4];
		p[i * 2 + 1] = HEX_CHARS[s[i] & 0x0f];
	}
	p[len * 2] = 0x00;
	return buf;
}

int signal_desc(int signal)
{
	sigset_t mask;
	sigemptyset(&mask);
	sigaddset(&mask, signal);

	if (sigprocmask(SIG_BLOCK, &mask, NULL) == -1) {
		PFATAL("sigprocmask(SIG_BLOCK, [%i])", signal);
	}

	int sfd = signalfd(-1, &mask, SFD_NONBLOCK);
	if (sfd == -1) {
		PFATAL("signalfd()");
	}
	return sfd;
}

const char **parse_argv(const char *str, char delim)
{
	int str_len = strlen(str);
	int i, items = 1;
	for (i = 0; i < str_len; i++) {
		if (str[i] == delim) {
			items += 1;
		}
	}

	char **argv = malloc(sizeof(char *) * (items + 1) + str_len + 1);
	char *nstr = (char *)&argv[items + 1];
	memcpy(nstr, str, str_len + 1);

	char delim_s[2] = {delim, '\x00'};
	char *s = nstr, *saveptr = NULL, **a = argv;

	for (;; s = NULL) {
		char *token = strtok_r(s, delim_s, &saveptr);
		if (token == NULL)
			break;

		a[0] = token;
		a += 1;
	}
	*a = NULL;

	return (const char **)argv;
}


src/main.c
// PMTUD
//
// Copyright (c) 2015 CloudFlare, Inc.

#include <errno.h>
#include <getopt.h>
#include <pcap.h>
#include <signal.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <unistd.h>
#include <arpa/inet.h>

#include "hashlimit.h"
#include "pmtudd.h"
#include "uevent.h"

#define IFACE_RATE_PPS 10.0
#define SRC_RATE_PPS 1.1

static void usage()
{
	fprintf(stderr,
		"Usage:\n"
		"\n"
		"    pmtudd [options]\n"
		"\n"
        "Path MTU Discovery Daemon listens for inbound IPv4 packets with the DF (Don't Fragment)\n"
        "bit set and sends back ICMP code 3 messages related to MTU detection\n"
		"\n"
		"Options:\n"
		"\n"
		"  --iface              Network interface to listen on\n"
		"  --src-rate           Pps limit from single source "
		"(default=%.1f pss)\n"
		"  --iface-rate         Pps limit to send on a single "
		"interface "
		"(default=%.1f pps)\n"
		"  --verbose            Print forwarded packets on screen\n"
		"  --dry-run            Don't inject packets, just dry run\n"
		"  --desired-mtu        The MTU to send back link\n"
		"  --cpu                Pin to particular cpu\n"
		"  --help               Print this message\n"
		"\n"
		"Example:\n"
		"\n"
		"    pmtudd --iface=eth2 --src-rate=%.1f --iface-rate=%.1f --desired-mtu=1420\n"
		"\n",
		SRC_RATE_PPS, IFACE_RATE_PPS, SRC_RATE_PPS, IFACE_RATE_PPS);
	exit(-1);
}

#define SNAPLEN 2048
#define BPF_FILTER "ip and ip[6] == 64 and greater %d"

static int on_signal(struct uevent *uevent, int sfd, int mask, void *userdata)
{
	volatile int *done = userdata;
	int buf[512];
	/* Drain. Socket should be NONBLOCK */
	int r = read(sfd, buf, sizeof(buf));
	if (r < 0) {
		PFATAL("read()");
	}

	*done = 1;
	return 0;
}

struct state
{
	pcap_t *pcap;
	int raw_sd;
	struct hashlimit *sources;
	struct hashlimit *ifaces;
	int verbose;
	int dry_run;
	uint16_t desired_mtu;
};

static int handle_packet(const uint8_t *p, unsigned data_len, void *userdata)
{
	struct state *state = userdata;

	const char *reason = "unknown";
	int l4_sport = -1;

	/* assumming DLT_EN10MB */

	/* 14 ethernet plus the maximum MTU */
	if (data_len < 14 + state->desired_mtu) {
		reason = "Packet size within bounds";
		goto reject;
	}

	/* Check whether we're being tricked, we never known */
	/*if (p[0] == p[6] && p[1] == p[7] && p[2] == p[8] && p[3] == p[9] &&
	    p[4] == p[10] && p[5] == p[11]) {
        reason = "Invalid Ethernet header";
        goto reject;
	}*/

	const uint8_t *hash = NULL;
	int hash_len = 0;

	/* Check whether the packet has a VLAN tag */
	unsigned l3_offset = 14;
	uint16_t eth_type = (((uint16_t)p[12]) << 8) | (uint16_t)p[13];
	if (eth_type == 0x8100) {
		eth_type = (((uint16_t)p[16]) << 8) | (uint16_t)p[17];
		l3_offset = 18;
	}

	int valid = 0;
    int l3_hdr_len = 0;

	/* Handle an IPv4 packet */
	if (eth_type == 0x0800 && (p[l3_offset] & 0xF0) == 0x40) {
		l3_hdr_len = (int)(p[l3_offset] & 0x0F) * 4;
		if (l3_hdr_len < 20 || l3_hdr_len > 24) {
			reason = "IPv4 header invalid length";
			goto reject;
		}

		/* Flags : R, DF, MF. We want DF = 1 and the others = 0 */
		if ((p[l3_offset + 6] & 0xE0) == 0x40) {
            valid = 1;
            hash = &p[l3_offset + 12];
            hash_len = 4;
		}
	}

    /* Handle an IPv6 packet : do not send anything back */
	if (eth_type == 0x86dd && (p[l3_offset] & 0xF0) == 0x60) {
		reason = "IPv6 is not supported";
        goto reject;
	}

	if (valid == 0 || hash == NULL || hash_len == 0) {
		reason = "Invalid IP packet or DF not set";
		goto reject;
	}

	/* Forge our new ICMP reply packet */
	struct icmp_packet *packet = malloc(sizeof(struct icmp_packet));

	/* Set the MAC src and dst addresses */
	int i;
	for (i = 0; i < 6; i++) {
        packet->eth_dst[i] = p[6 + i];
	}

	for (i = 0; i < 6; i++) {
        packet->eth_src[i] = p[i];
	}
	packet->eth_type = 0x0008;

	uint16_t ip_chksum = 0;

	packet->ip_v_ihl = (0x04 << 4) | 0x05;
	packet->ip_tos = 0x00;
    ip_chksum += (packet->ip_v_ihl << 8) | packet->ip_tos;

	packet->ip_id = 0x0000;
	packet->ip_flags_offset = 0x0000;

	packet->ip_ttl = (u_int8_t )0x40;
	packet->ip_proto = (u_int8_t )0x01;
    ip_chksum += (packet->ip_ttl << 8) | packet->ip_proto;

	packet->ip_chksum = 0x0000;

	/* Set the IP src and dst addresses */
    for (i = 0; i < 4; i ++) {
        packet->ip_src_addr[i] = p[l3_offset + 12 + 4 + i];
    }
    ip_chksum += add_checksum(packet->ip_src_addr, 4);

	for (i = 0; i < 4; i ++) {
        packet->ip_dst_addr[i] = p[l3_offset + 12 + i];
	}
    ip_chksum += add_checksum(packet->ip_dst_addr, 4);

    uint16_t icmp_chksum = 0;

    packet->icmp_type = (u_int8_t )0x3;
    packet->icmp_code = (u_int8_t )0x4;
    icmp_chksum += ((uint16_t)packet->icmp_type << 8) | (uint16_t)packet->icmp_code;

    packet->icmp_chksum = 0x0000;
    packet->icmp_unused = 0x0000;
    packet->icmp_mtu = htons(state->desired_mtu);
    icmp_chksum += (uint16_t)state->desired_mtu;

    /* Copy the IP header + the first 8 bytes of the IP packet */
    for (i = 0; i < l3_hdr_len; i++) {
        packet->icmp_data[i] = p[l3_offset + i];
    }

    for (i = 0; i < 8; i++) {
        packet->icmp_data[i + l3_hdr_len] = p[l3_offset + l3_hdr_len + i];
    }
    icmp_chksum += add_checksum(packet->icmp_data, l3_hdr_len + 8);

    /* Set the IP packet length */
    /* ip + icmp + original_ip + 8 bytes of data */
    packet->ip_length = htons((u_int16_t)36 + (u_int16_t)l3_hdr_len);
    ip_chksum += ntohs(packet->ip_length);

    /* Compute the ICMP and IP checksums */
    packet->icmp_chksum = (icmp_chksum >> 16) + (icmp_chksum & 0xffff);
    packet->icmp_chksum += (packet->icmp_chksum >> 16);
    packet->icmp_chksum = htons(~packet->icmp_chksum - 4);

    packet->ip_chksum = (ip_chksum >> 16) + (ip_chksum & 0xffff);
    packet->ip_chksum += (packet->ip_chksum >> 16);
    packet->ip_chksum = htons(~packet->ip_chksum - 2);

	/* Check if the limits will be reached */
	int limit_src = hashlimit_check_hash(state->sources, hash, hash_len);
	int limit_iface = hashlimit_check(state->ifaces, 0);

	if (limit_src == 0) {
		reason = "Ratelimited on source IP";
		goto reject;
	}
	if (limit_iface == 0) {
		reason = "Ratelimited on outgoing interface";
		goto reject;
	}

	hashlimit_subtract_hash(state->sources, hash, hash_len);
	hashlimit_subtract(state->ifaces, 0);

	u_int8_t *pp;
	build_packet(packet, &pp, l3_hdr_len + 8, ntohs(packet->ip_length) + 14);

	reason = "transmitting";
	if (state->verbose > 2) {
		printf("%s %s mtu=%i sport=%i  data=%s\n",
		       ip_to_string(hash, hash_len), reason, state->desired_mtu,
		       l4_sport, to_hex(pp, ntohs(packet->ip_length) + 14));
	} else if (state->verbose) {
		printf("%s %s mtu=%i sport=%i\n", ip_to_string(hash, hash_len),
		       reason, state->desired_mtu, l4_sport);
	}

	if (state->dry_run == 0) {
		int r = send(state->raw_sd, pp, ntohs(packet->ip_length) + 14, 0);
        //int r = send(state->raw_sd, p, data_len, 0);
		/* ENOBUFS happens during IRQ storms okay to ignore */
		if (r < 0 && errno != ENOBUFS) {
			PFATAL("send()");
		}
	}
	return 1;

reject:
	if (state->verbose > 2) {
		printf("%s %s mtu=%i sport=%i  %s\n",
		       ip_to_string(hash, hash_len), reason, state->desired_mtu,
		       l4_sport, to_hex(p, data_len));
	} else if (state->verbose > 1) {
		printf("%s %s mtu=%i sport=%i\n", ip_to_string(hash, hash_len),
		       reason, state->desired_mtu, l4_sport);
	}

	return -1;
}

static int handle_pcap(struct uevent *uevent, int sfd, int mask, void *userdata)
{
	struct state *state = userdata;

	while (1) {
		struct pcap_pkthdr *hdr;
		const uint8_t *data;

		int r = pcap_next_ex(state->pcap, &hdr, &data);

		switch (r) {
		case 1:
			if (hdr->len == hdr->caplen) {
				handle_packet(data, hdr->caplen, state);
			} else {
				/* Partial caputre */
			}
			break;

		case 0:
			/* Timeout */
			return 0;

		case -1:
			FATAL("pcap_next_ex(): %s", pcap_geterr(state->pcap));
			break;

		case -2:
			return 0;
		}
	}
}

int main(int argc, char *argv[])
{
	static struct option long_options[] = {
		{"iface", required_argument, 0, 'i'},
		{"src-rate", required_argument, 0, 's'},
		{"iface-rate", required_argument, 0, 'r'},
		{"verbose", no_argument, 0, 'v'},
		{"dry-run", no_argument, 0, 'd'},
		{"cpu", required_argument, 0, 'c'},
		{"help", no_argument, 0, 'h'},
		{"desired-mtu", required_argument, 0, 'm'},
		{NULL, 0, 0, 0}};

	const char *optstring = optstring_from_long_options(long_options);
	const char *iface = NULL;

	double src_rate = SRC_RATE_PPS;
	double iface_rate = IFACE_RATE_PPS;
	int verbose = 0;
	int dry_run = 0;
	int taskset_cpu = -1;
	int desired_mtu = -1;

	optind = 1;
	while (1) {
		int option_index = 0;
		int arg = getopt_long(argc, argv, optstring, long_options,
				      &option_index);
		if (arg == -1) {
			break;
		}

		switch (arg) {
		case 0:
			FATAL("Unknown option: %s", argv[optind]);
			break;

		case 'h':
			usage();
			break;

		case '?':
			exit(-1);
			break;

		case 'i':
			iface = optarg;
			break;

		case 's':
			src_rate = atof(optarg);
			if (src_rate <= 0.0) {
				FATAL("Rates must be greater than zero");
			}
			break;

		case 'r':
			iface_rate = atof(optarg);
			if (iface_rate <= 0.0) {
				FATAL("Rates must be greater than zero");
			}
			break;
		case 'm':
			desired_mtu = atoi(optarg);
			if (desired_mtu < 0 || desired_mtu > 65535) {
				FATAL("MTU must be within range "
				      "0..65535");
			}
			break;

		case 'v':
			verbose++;
			break;

		case 'd':
			dry_run = 1;
			break;

		case 'c':
			taskset_cpu = atoi(optarg);
			break;

		default:
			FATAL("Unknown option %c: %s", arg,
			      str_quote(argv[optind]));
		}
	}

	if (argv[optind]) {
		FATAL("Not sure what you mean by %s", str_quote(argv[optind]));
	}

	if (iface == NULL) {
		FATAL("Specify interface with --iface option");
	}

	if (desired_mtu == -1) {
        FATAL("Specify desired MTU with --desired-mtu option");
	}

	if (set_core_dump(1) < 0) {
		ERRORF("[ ] Failed to enable core dumps, continuing anyway.\n");
	}

	if (taskset_cpu > -1) {
		if (taskset(taskset_cpu)) {
			ERRORF("[ ] sched_setaffinity(%i): %s\n", taskset_cpu,
			       strerror(errno));
		}
	}

	struct pcap_stat stats = {0, 0, 0};
	struct state state;
	memset(&state, 0, sizeof(struct state));
	state.sources = hashlimit_alloc(8191, src_rate, src_rate * 1.9);
	state.ifaces = hashlimit_alloc(32, iface_rate, iface_rate * 1.9);
	state.verbose = verbose;
	state.dry_run = dry_run;
	state.desired_mtu = desired_mtu;
	state.raw_sd = setup_raw(iface);

	struct uevent uevent;
	uevent_new(&uevent);


    char * filter = (char *)malloc(sizeof(BPF_FILTER) + 3* sizeof(char));

    sprintf(filter, BPF_FILTER, state.desired_mtu);
    state.pcap = setup_pcap(iface, filter, SNAPLEN, &stats);
    int pcap_fd = pcap_get_selectable_fd(state.pcap);
    if (pcap_fd < 0) {
        PFATAL("pcap_get_selectable_fd()");
    }
    uevent_yield(&uevent, pcap_fd, UEVENT_READ, handle_pcap,
             &state);


	volatile int done = 0;
	uevent_yield(&uevent, signal_desc(SIGINT), UEVENT_READ, on_signal,
		     (void *)&done);
	uevent_yield(&uevent, signal_desc(SIGTERM), UEVENT_READ, on_signal,
		     (void *)&done);

	fprintf(stderr, "[*] #%i Started pmtudd ", getpid());
	fprintf(stderr, "pcap on iface=%s ", str_quote(iface));

	fprintf(stderr,
		"rates={iface=%.1f pps source=%.1f pps}, verbose=%i, "
		"dry_run=%i\n",
		iface_rate, src_rate, verbose, dry_run);

	while (done == 0) {
		struct timeval timeout =
			NSEC_TIMEVAL(MSEC_NSEC(24 * 60 * 60 * 1000UL));
		int r = uevent_select(&uevent, &timeout);
		if (r != 0) {
			continue;
		}
	}
	fprintf(stderr, "[*] #%i Quitting\n", getpid());

	unsetup_pcap(state.pcap, iface, &stats);
	fprintf(stderr, "[*] #%i recv=%i drop=%i ifdrop=%i\n", getpid(),
		stats.ps_recv, stats.ps_drop, stats.ps_ifdrop);

	close(state.raw_sd);

	hashlimit_free(state.sources);
	hashlimit_free(state.ifaces);

	return 0;
}
